# Matthew Peetz
# Regis University
# Week 3 Lab: Working with PDFs

This week you will be retrieving and cleansing data from a survey generated by the National Center for Immunization and Respiratory Diseases about National Immunizations in Children. In completing this assignment, you can combine topics discussed in several of our prior lectures.

File needed to complete this assignment are located in the assign3_data folder:
* NISPUF14_CODEBOOK.PDF
* nispuf14.dat

# Assignment Requirements

Here this week's assingment you will need to complete the following:

* Retrieve all of the data within nispuf14.dat and store it in a more <i> accessible format</i>
* <i> Accessible format </i> can be any of the following:
    - csv file
    - json file
    - relational database
* For this assignment, feel free to use a dataframe (python library Pandas) for intermediate steps. 

# Deliverables

Upload your Jupyter Notebook, along with any files that you used and were not provided to you. This includes your final output and any intermediate files that you might create.

**Note::** Make sure you have clearly indicated each assignment requirement within your notebook and include your thoughts/process/analysis as you go. Code alone is not enough.

<div class="alert alert-block alert-danger">
<b>Important:</b> Using Pandas read_fwf() function, or something like it, to parse the nispuf14.dat will result in a <u>zero</u> on this assignment. The purpose of this exercise is for you understand how to read a PDF that includes a table, parse the tabular data, and systematically parse the data file. 
</div>

# Helpful Information for the Assignment

## What's in these two files?

I'm glad you asked that! And to be honest, you probably will not like the answer.

NISPUF14_CODEBOOK.PDF is a PDF that contains a description of the format for the data in nispuf14.dat. In other words, the PDF tells you how to read the data in nispuf14.dat.

Why would we need a PDF to tell us how to read our data?  Well, this data file is stored in a positional format. This means that each character's value and relative position provides meaning within the dataset.

Here's what the data in nispuf14.dat looks like.
<img align="left" style="padding-right:10px;" src="figures_wk3/positional_data.jpeg" width = 800><br>

Ugly? Yes! And very much so. However, data in this format is not all that uncommon. Mainframe computers operate on positional formating. 

Q - Who still uses mainframe computers?<br>
A - Mainframes are more prevalent than you'd think. Any industry with a large volume of daily mathematical calculations will most likely use a mainframe computer as part of their everyday operation (i.e.: the banking industry) Indeed, the website and customer-facing applications are not run on a mainframe computer, but the nightly accounting processes probably are. 

The following [article](https://www.thocp.net/hardware/mainframe.htm) describes the mainframe computer's history and how it has evolved over the years.


## How are we supposed to read that?

This is where NISPUF14_CODEBOOK.PDF comes into the picture. Section 1 of the PDF describes the positional formatting information for each data field. Here's how it works!

As an example, let's say that our data file looked like this:<br>
CAT  FLUFFY410<br>
DOG  FIDO  522<br>
BIRD CHIRP 2 1<br>

At a glance, we can determine that each line contains information about animals. We can see a field representing an animal_type and perhaps an animal_name.  However, we have little to no information about the numerics at each line's end. Or even how many fields the numeric group is representing. The last line leads us to believe that more than one field might be defined, but we are not confident at this point.

## Does this come with a 'Magical Decoder Ring'?

Short of an actual magical ring, I'd settle for a description of each field and its relative position in the line.  It would be even better if the report were written for future reference.

Let's look at the above animal dataset in conjunction with the  following description:<br>
Type 1 5<br>
Name 6 11<br>
Age 12 12<br>
Weight 13 14<br>

Aaahhhhh! Now everything is starting to come together! We can confirm that the first field is indeed animal_type, and the second is animal_name. However, we now know that the numeric grouping is two fields, animal_age and animal_weight. We can also see that animal_age is a single digit, and animal_weight is a 2-digit numeric. We can also determine at this point that the animal_name on the first line is 'FLUFFY' and not 'FLUFFY410'.

Time to add a little code to this example.

In [1]:
# Load the sample data into a list
animal_data = ['CAT  FLUFFY410', 'DOG  FIDO  522', 'BIRD CHIRP 2 1' ]

# processing each animal_data line
for  line in animal_data:
    animal_type = line[0:5]
    animal_name = line[5:11]
    age = line[11]
    weight = line[12:14]
    
    print(f'{animal_name} is a {animal_type} that is {age} years old and weights {weight} pound(s)')


FLUFFY is a CAT   that is 4 years old and weights 10 pound(s)
FIDO   is a DOG   that is 5 years old and weights 22 pound(s)
CHIRP  is a BIRD  that is 2 years old and weights  1 pound(s)


Hopefully, things are looking less scary at this point? 

Retrieving data from a positionally formatted file is just chunking the larger string into smaller pieces. The trick is in determining where to make those chunks. 

The key to all this is the 'magical decoder' description because there are no other clues in the file itself. Unlike a CSV-type file, positional formatted files don't have a delimited to help identify individual data elements. 

That being said, positional formats account for every character within a row, meaning that even unused characters are given a value. In our example above, a blank character(' ') was used to fill unused characters. The value used to represent unused characters can be anything. For example, if '-' was used instead of a ', 'our sample data would have looked like this:

CAT--FLUFFY410<br>
DOG--FIDO--522<br>
BIRD-CHIRP-2-1<br>

Let's see if our code above will still work?

In [2]:
# Load the sample data into a list
animal_data2 = ['CAT--FLUFFY410', 'DOG--FIDO--522', 'BIRD-CHIRP-2-1' ]

# processing each animal_data line
for  line in animal_data2:
    animal_type = line[0:5]
    animal_name = line[5:11]
    age = line[11]
    weight = line[12:14]
    
    print(f'{animal_name} is a {animal_type} that is {age} years old and weights {weight} pound(s)')

FLUFFY is a CAT-- that is 4 years old and weights 10 pound(s)
FIDO-- is a DOG-- that is 5 years old and weights 22 pound(s)
CHIRP- is a BIRD- that is 2 years old and weights -1 pound(s)


Aside from changing our dataset's initial list, no coding changes were needed. 

Our output looks a little different, but that's because of the other unused character representations. The above examples have their respective new character values in the data elements.  It's just easier to see in the second example over the first.

Let's try stripping out the unused characters in both examples.

In [3]:
# working with the second dataset, animal_data2, first.

# processing each animal_data line
for  line in animal_data2:
    animal_type = line[0:5].strip('-')
    animal_name = line[5:11].strip('-')
    age = line[11].strip('-')
    weight = line[12:14].strip('-')
    
    print(f'{animal_name} is a {animal_type} that is {age} years old and weights {weight} pound(s)')

FLUFFY is a CAT that is 4 years old and weights 10 pound(s)
FIDO is a DOG that is 5 years old and weights 22 pound(s)
CHIRP is a BIRD that is 2 years old and weights 1 pound(s)


In [4]:
# repeat the same things with the first set, animal_data.

# processing each animal_data line
for  line in animal_data:
    animal_type = line[0:5].strip('-')
    animal_name = line[5:11].strip('-')
    age = line[11].strip('-')
    weight = line[12:14].strip('-')
    
    print(f'{animal_name} is a {animal_type} that is {age} years old and weights {weight} pound(s)')

FLUFFY is a CAT   that is 4 years old and weights 10 pound(s)
FIDO   is a DOG   that is 5 years old and weights 22 pound(s)
CHIRP  is a BIRD  that is 2 years old and weights  1 pound(s)


<div class="alert alert-success">
Success!! The two outputs match!
</div>

## Back to our assignment

Section 1 of NISPUF14_CODEBOOK.PDF contains the description of the positional format for nispuf14.dat.

<div class="alert alert-block alert-info">
<b>Helpful Hint::</b> Combining pyPDF2 and Tabula would work great for parsing the information within section 1 of NISPUF14_CODEBOOK.PDF. pyPDF2 to retrieve section 1 of the PDF and Tabula for getting the positional formatting information off the PDF and into a pandas dataframe.
</div>

Installation reminders from our week3 lecture notebook.
<div class="alert alert-block alert-success">
<b>Installation - PyPDF2::</b> PyPDF2 can be installed as normal using pip.
</div>

<div class="alert alert-block alert-success">
<b>Installation - Tabula::</b> To install the tabula package, you can use pip as shown before. https://pypi.org/project/tabula-py/
</div>

<div class="alert alert-block alert-success">
<b>Installation - Java::</b> Note: in order to use tabula, you need to have the latest version of java installed. https://aegis4048.github.io/parse-pdf-files-while-retaining-structure-with-tabula-py has some useful information if you need help getting java installed on your machine.
</div>

## Assignment Approach

<div class="alert alert-block alert-warning">
<b>One possible solution: </b> Students are encouraged to define their approach when completing any assignment in this class.  Below, I have shared my approach to the assignment for this week.  Feel free to use some or all of this design, if you'd like.
</div>

for each line in the file:

    data_line = new list
    for each variable (line) found in the dataframe:
        create a dictionary with variable name as key, 
        use start / end position numbers as a slice to give the dictionary's value
        append dictionary to data_line
    write data_line to CSV file


* Retrieve all of the data within nispuf14.dat and store it in a more <i> accessible format</i>

Combining pyPDF2 and Tabula would work great for parsing the information within section 1 of NISPUF14_CODEBOOK.PDF. pyPDF2 to retrieve section 1 of the PDF and Tabula for getting the positional formatting information off the PDF and into a pandas dataframe.

### Strategy 
* First I am going to read in the column and spacing catagories which are located on pages 5-17 of the PDF
* I will then extract the headers
* I will use tabula to pull information on where to slice the main frame data
* Next I will turn all the rows into a dicionary, and eventually two lists, so that they can accessed and sliced based on the notation provided in the PDF and provide the header information
* Next I will insert commas at the slice points into the lists
* Finally I will create a json file using the headers and the comma inserted lists

In [79]:
#import PyPDF2 
from PyPDF2 import PdfReader, PdfWriter,PdfMerger

# create a file object for our PDF
pdfFileObj = open('assign_wk3/NISPUF14_CODEBOOK.PDF', 'rb') 


# create an object to access the PDF
reader = PdfReader(pdfFileObj) 

Taking a look at the length of the PDF

In [80]:
print(len(reader.pages)) 

250


In [81]:
# import tabula library
import tabula

Reading in the data from the pages that contain the information that is needed, pages 5-21

In [82]:
# pdf location
pdf_file = "assign_wk3/NISPUF14_CODEBOOK.PDF"

# for the pages argument you can specify a single paper, a range of pages, or 'all'.
# I know my PDF only has one page to it, so 1 or 'all' will work here.
table_data = tabula.read_pdf(pdf_file, pages = '5-21')

In [84]:
table_data

[    Variable Name  Position  Position.1  Section   
 0         SEQNUMC       1.0         6.0      1.0  \
 1        SEQNUMHH       7.0        11.0      1.0   
 2            PDAT      12.0        12.0      1.0   
 3        PROVWT_D      13.0        31.0      1.0   
 4             NaN       NaN         NaN      NaN   
 5   PROVWT_D_TERR      32.0        50.0      1.0   
 6             NaN       NaN         NaN      NaN   
 7         RDDWT_D      51.0        69.0      1.0   
 8             NaN       NaN         NaN      NaN   
 9    RDDWT_D_TERR      70.0        88.0      1.0   
 10            NaN       NaN         NaN      NaN   
 11        STRATUM      89.0        92.0      1.0   
 12           YEAR      93.0        96.0      1.0   
 13       AGECPOXR      97.0        97.0      2.0   
 14       HAD_CPOX      98.0        99.0      2.0   
 15       SHOTCARD     100.0       100.0      2.0   
 16         AGEGRP     101.0       101.0      3.0   
 17      BF_ENDR06     102.0       109.0      

Creating a pandas data frame of the information from tablua

In [85]:
import pandas as pd
import numpy as np

i = 0
new_df = pd.DataFrame()

while i < 17:
    df1 = table_data[i]
    new_df = pd.concat([new_df,df1])
    i = i+1

Taking a look at the data frame

In [86]:
new_df

Variable Name  Position  Position.1  Section   
0        SEQNUMC       1.0         6.0      1.0  \
1       SEQNUMHH       7.0        11.0      1.0   
2           PDAT      12.0        12.0      1.0   
3       PROVWT_D      13.0        31.0      1.0   
4            NaN       NaN         NaN      NaN   
..           ...       ...         ...      ...   
8        INS_4_5     861.0       862.0     10.0   
9            NaN       NaN         NaN      NaN   
10         INS_6     863.0       864.0     10.0   
11           NaN       NaN         NaN      NaN   
12        INS_11     865.0       866.0     10.0   

                                       Variable Label  
0                             UNIQUE CHILD IDENTIFIER  
1                         UNIQUE HOUSEHOLD IDENTIFIER  
2                    CHILD HAS ADEQUATE PROVIDER DATA  
3    FINAL DUAL-FRAME PROVIDER-PHASE WEIGHT (EXCLUDES  
4                                        TERRITORIES)  
..                                                ...  
8   IS CHILD COVERED BY INDIAN HEALTH SERVICE, MIL...  
9                CARE, TRICARE, CHAMPUS, OR CHAMP-VA?  
10  IS CHILD COVERED BY ANY OTHER HEALTH INSURANCE...  
11                                         CARE PLAN?  
12  ANY TIME WHEN CHILD WAS NOT COVERED BY ANY HEALTH  

[693 rows x 5 columns]

In [87]:
new_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 693 entries, 0 to 12
Data columns (total 5 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Variable Name   461 non-null    object 
 1   Position        461 non-null    float64
 2   Position.1      461 non-null    float64
 3   Section         461 non-null    float64
 4   Variable Label  693 non-null    object 
dtypes: float64(3), object(2)
memory usage: 32.5+ KB


In [88]:
type(new_df)

pandas.core.frame.DataFrame

In [89]:
new_df.tail(10)

Variable Name  Position  Position.1  Section   
3          INS_1     853.0       854.0     10.0  \
4            NaN       NaN         NaN      NaN   
5          INS_2     855.0       856.0     10.0   
6          INS_3     857.0       858.0     10.0   
7         INS_3A     859.0       860.0     10.0   
8        INS_4_5     861.0       862.0     10.0   
9            NaN       NaN         NaN      NaN   
10         INS_6     863.0       864.0     10.0   
11           NaN       NaN         NaN      NaN   
12        INS_11     865.0       866.0     10.0   

                                       Variable Label  
3   IS CHILD COVERED BY HEALTH INSURANCE PROVIDED ...  
4                                  EMPLOYER OR UNION?  
5              IS CHILD COVERED BY ANY MEDICAID PLAN?  
6                         IS CHILD COVERED BY S-CHIP?  
7    IS CHILD COVERED BY ANY MEDICAID PLAN OR S-CHIP?  
8   IS CHILD COVERED BY INDIAN HEALTH SERVICE, MIL...  
9                CARE, TRICARE, CHAMPUS, OR CHAMP-VA?  
10  IS CHILD COVERED BY ANY OTHER HEALTH INSURANCE...  
11                                         CARE PLAN?  
12  ANY TIME WHEN CHILD WAS NOT COVERED BY ANY HEALTH

There are several columns that are not needed, they can be dropped

In [90]:
small_df = new_df.drop(['Section','Variable Label'], axis = 1)

In [91]:
small_df.head(5)

Variable Name  Position  Position.1
0       SEQNUMC       1.0         6.0
1      SEQNUMHH       7.0        11.0
2          PDAT      12.0        12.0
3      PROVWT_D      13.0        31.0
4           NaN       NaN         NaN

Setting the index as the variable name

In [92]:
small_df = small_df.set_index('Variable Name')

Checking to make sure that that worked

In [93]:
small_df.head(5)

Position  Position.1
Variable Name                      
SEQNUMC             1.0         6.0
SEQNUMHH            7.0        11.0
PDAT               12.0        12.0
PROVWT_D           13.0        31.0
NaN                 NaN         NaN

I have a lot of rows that contain "NaN", if I can change the whole file to be numeric I can then drop them using a pandas command

In [94]:
# Changing all the values to numeric
number_df = small_df.apply(pd.to_numeric, errors='coerce')
number_df

Position  Position.1
Variable Name                      
SEQNUMC             1.0         6.0
SEQNUMHH            7.0        11.0
PDAT               12.0        12.0
PROVWT_D           13.0        31.0
NaN                 NaN         NaN
...                 ...         ...
INS_4_5           861.0       862.0
NaN                 NaN         NaN
INS_6             863.0       864.0
NaN                 NaN         NaN
INS_11            865.0       866.0

[693 rows x 2 columns]

In [95]:
# Dropping the NaN values
full_df = number_df.dropna()
full_df.head(10)

Position  Position.1
Variable Name                      
SEQNUMC             1.0         6.0
SEQNUMHH            7.0        11.0
PDAT               12.0        12.0
PROVWT_D           13.0        31.0
PROVWT_D_TERR      32.0        50.0
RDDWT_D            51.0        69.0
RDDWT_D_TERR       70.0        88.0
STRATUM            89.0        92.0
YEAR               93.0        96.0
AGECPOXR           97.0        97.0

I can now create a dictionary from the data frame in order to access the individuals columns

In [96]:
slicer = full_df.to_dict('index')

Checking to see what was created

In [97]:
slicer

{'SEQNUMC': {'Position': 1.0, 'Position.1': 6.0},
 'SEQNUMHH': {'Position': 7.0, 'Position.1': 11.0},
 'PDAT': {'Position': 12.0, 'Position.1': 12.0},
 'PROVWT_D': {'Position': 13.0, 'Position.1': 31.0},
 'PROVWT_D_TERR': {'Position': 32.0, 'Position.1': 50.0},
 'RDDWT_D': {'Position': 51.0, 'Position.1': 69.0},
 'RDDWT_D_TERR': {'Position': 70.0, 'Position.1': 88.0},
 'STRATUM': {'Position': 89.0, 'Position.1': 92.0},
 'YEAR': {'Position': 93.0, 'Position.1': 96.0},
 'AGECPOXR': {'Position': 97.0, 'Position.1': 97.0},
 'HAD_CPOX': {'Position': 98.0, 'Position.1': 99.0},
 'SHOTCARD': {'Position': 100.0, 'Position.1': 100.0},
 'AGEGRP': {'Position': 101.0, 'Position.1': 101.0},
 'BF_ENDR06': {'Position': 102.0, 'Position.1': 109.0},
 'BF_EXCLR06': {'Position': 110.0, 'Position.1': 117.0},
 'BF_FORMR08': {'Position': 118.0, 'Position.1': 125.0},
 'BFENDFL06': {'Position': 126.0, 'Position.1': 126.0},
 'BFFORMFL06': {'Position': 127.0, 'Position.1': 127.0},
 'C1R': {'Position': 128.0, 'Po

Taking a look at all of the potential header values

In [98]:
print(slicer.keys())

dict_keys(['SEQNUMC', 'SEQNUMHH', 'PDAT', 'PROVWT_D', 'PROVWT_D_TERR', 'RDDWT_D', 'RDDWT_D_TERR', 'STRATUM', 'YEAR', 'AGECPOXR', 'HAD_CPOX', 'SHOTCARD', 'AGEGRP', 'BF_ENDR06', 'BF_EXCLR06', 'BF_FORMR08', 'BFENDFL06', 'BFFORMFL06', 'C1R', 'C5R', 'CBF_01', 'CEN_REG', 'CHILDNM', 'CWIC_01', 'CWIC_02', 'EDUC1', 'FRSTBRN', 'I_HISP_K', 'INCPORAR', 'INCPOV1', 'INCQ298A', 'INTRP', 'LANGUAGE', 'M_AGEGRP', 'MARITAL2', 'MOBIL_I', 'NUM_PHONE', 'NUM_CELLS_HH', 'NUM_CELLS_PARENTS', 'RACE_K', 'RACEETHK', 'RENT_OWN', 'SEX', 'ESTIAP14', 'EST_GRANT', 'STATE', 'D6R', 'D7', 'N_PRVR', 'PROV_FAC', 'REGISTRY', 'VFC_ORDER', 'HEP_BRTH', 'HEP_FLAG', 'P_NUHEPX', 'P_NUHIBX', 'P_NUHPHB', 'P_NUMDAH', 'P_NUMDHI', 'P_NUMDIH', 'P_NUMDTA', 'P_NUMDTP', 'P_NUMFLU', 'P_NUMFLUL', 'P_NUMFLUM', 'P_NUMFLUN', 'P_NUMHEA', 'P_NUMHEN', 'P_NUMHEP', 'P_NUMHG', 'P_NUMHHY', 'P_NUMHIB', 'P_NUMHIN', 'P_NUMHION', 'P_NUMHM', 'P_NUMHS', 'P_NUMIPV', 'P_NUMMCN', 'P_NUMMMR', 'P_NUMMMRX', 'P_NUMMMX', 'P_NUMMP', 'P_NUMMPR', 'P_NUMMRV', 'P_NUMMS

Taking a look at the possible header values

In [99]:
headers = list(slicer.keys())
print(len(headers))

461


I now have the header values. I need to figure out where to inset commas in the rest of the data, first I will take a look at all the values in the dictionary. 

In [100]:
print(slicer.values())

dict_values([{'Position': 1.0, 'Position.1': 6.0}, {'Position': 7.0, 'Position.1': 11.0}, {'Position': 12.0, 'Position.1': 12.0}, {'Position': 13.0, 'Position.1': 31.0}, {'Position': 32.0, 'Position.1': 50.0}, {'Position': 51.0, 'Position.1': 69.0}, {'Position': 70.0, 'Position.1': 88.0}, {'Position': 89.0, 'Position.1': 92.0}, {'Position': 93.0, 'Position.1': 96.0}, {'Position': 97.0, 'Position.1': 97.0}, {'Position': 98.0, 'Position.1': 99.0}, {'Position': 100.0, 'Position.1': 100.0}, {'Position': 101.0, 'Position.1': 101.0}, {'Position': 102.0, 'Position.1': 109.0}, {'Position': 110.0, 'Position.1': 117.0}, {'Position': 118.0, 'Position.1': 125.0}, {'Position': 126.0, 'Position.1': 126.0}, {'Position': 127.0, 'Position.1': 127.0}, {'Position': 128.0, 'Position.1': 128.0}, {'Position': 129.0, 'Position.1': 130.0}, {'Position': 131.0, 'Position.1': 132.0}, {'Position': 133.0, 'Position.1': 133.0}, {'Position': 134.0, 'Position.1': 134.0}, {'Position': 135.0, 'Position.1': 136.0}, {'Po

That is kind of a mess, in particular the position value has a number in it which is going to mess up my eventual plan to extract all the numbers. So I am going to change it to, "start" and "end"

In [105]:
string = str(slicer.values())
new_string = string.replace("Position.1", "end")
new_string = new_string.replace("Position", 'start')
new_string

"dict_values([{'start': 1.0, 'end': 6.0}, {'start': 7.0, 'end': 11.0}, {'start': 12.0, 'end': 12.0}, {'start': 13.0, 'end': 31.0}, {'start': 32.0, 'end': 50.0}, {'start': 51.0, 'end': 69.0}, {'start': 70.0, 'end': 88.0}, {'start': 89.0, 'end': 92.0}, {'start': 93.0, 'end': 96.0}, {'start': 97.0, 'end': 97.0}, {'start': 98.0, 'end': 99.0}, {'start': 100.0, 'end': 100.0}, {'start': 101.0, 'end': 101.0}, {'start': 102.0, 'end': 109.0}, {'start': 110.0, 'end': 117.0}, {'start': 118.0, 'end': 125.0}, {'start': 126.0, 'end': 126.0}, {'start': 127.0, 'end': 127.0}, {'start': 128.0, 'end': 128.0}, {'start': 129.0, 'end': 130.0}, {'start': 131.0, 'end': 132.0}, {'start': 133.0, 'end': 133.0}, {'start': 134.0, 'end': 134.0}, {'start': 135.0, 'end': 136.0}, {'start': 137.0, 'end': 138.0}, {'start': 139.0, 'end': 139.0}, {'start': 140.0, 'end': 140.0}, {'start': 141.0, 'end': 141.0}, {'start': 142.0, 'end': 157.0}, {'start': 158.0, 'end': 158.0}, {'start': 159.0, 'end': 160.0}, {'start': 161.0, 'e

I can now strip out all the numbers from the sheet and create a list of them

In [107]:
# The re (regular expression) library will help with this
import re

numbers = re.findall(r'\d+', new_string)
print(numbers)

['1', '0', '6', '0', '7', '0', '11', '0', '12', '0', '12', '0', '13', '0', '31', '0', '32', '0', '50', '0', '51', '0', '69', '0', '70', '0', '88', '0', '89', '0', '92', '0', '93', '0', '96', '0', '97', '0', '97', '0', '98', '0', '99', '0', '100', '0', '100', '0', '101', '0', '101', '0', '102', '0', '109', '0', '110', '0', '117', '0', '118', '0', '125', '0', '126', '0', '126', '0', '127', '0', '127', '0', '128', '0', '128', '0', '129', '0', '130', '0', '131', '0', '132', '0', '133', '0', '133', '0', '134', '0', '134', '0', '135', '0', '136', '0', '137', '0', '138', '0', '139', '0', '139', '0', '140', '0', '140', '0', '141', '0', '141', '0', '142', '0', '157', '0', '158', '0', '158', '0', '159', '0', '160', '0', '161', '0', '161', '0', '162', '0', '162', '0', '163', '0', '163', '0', '164', '0', '164', '0', '165', '0', '165', '0', '166', '0', '167', '0', '168', '0', '169', '0', '170', '0', '171', '0', '172', '0', '172', '0', '173', '0', '173', '0', '174', '0', '175', '0', '176', '0', '176

I now need a way to get rid of all the zeros that were created from the decimal points, I found the below method for doing so on geeks for geeks

In [108]:
# https://www.geeksforgeeks.org/remove-all-the-occurrences-of-an-element-from-a-list-in-python/
def remove_items(test_list, item):
 
    # using list comprehension to perform the task
    res = [i for i in test_list if i != item]
 
    return res

In [109]:
item = '0'

results = remove_items(numbers, item)

print(results)

['1', '6', '7', '11', '12', '12', '13', '31', '32', '50', '51', '69', '70', '88', '89', '92', '93', '96', '97', '97', '98', '99', '100', '100', '101', '101', '102', '109', '110', '117', '118', '125', '126', '126', '127', '127', '128', '128', '129', '130', '131', '132', '133', '133', '134', '134', '135', '136', '137', '138', '139', '139', '140', '140', '141', '141', '142', '157', '158', '158', '159', '160', '161', '161', '162', '162', '163', '163', '164', '164', '165', '165', '166', '167', '168', '169', '170', '171', '172', '172', '173', '173', '174', '175', '176', '176', '177', '179', '180', '181', '182', '183', '184', '184', '185', '185', '186', '186', '187', '187', '188', '188', '189', '189', '190', '190', '191', '191', '192', '192', '193', '193', '194', '194', '195', '195', '196', '196', '197', '197', '198', '198', '199', '199', '200', '200', '201', '201', '202', '202', '203', '203', '204', '204', '205', '205', '206', '206', '207', '207', '208', '208', '209', '209', '210', '210', '2

Taking a look at the list, there should be 461 pairs of numbers

In [111]:
print(len(results))

922


That is awesome! Right where everything needs to be to use those values to insert commas, now it is time to pull in the main frame file.

In [113]:
main_frame = open('assign_wk3/nispuf14.dat')
main_frame.readlines(1000)

['000011000012    .                  .               218.30024855484000 218.3002485548400010222014. 223365.2500152.1875182.6250..3 2 131 2 .4223.000000000000001142131299999912 12 222212..0..........................................................................................   .   .   .   .   .   ....   .   .   .   .   .   ....   .   .   .   ......   .   .   .   .   .  .  ...   .   .   .   .   .  .  ...   .   .  .  ...............  .........   .   .   .   .   .  .  ...   .   .   .   .  .  ....  .........  .  .   .  .  .....   .   .   ....... . . . . . .... . . . . . .... . . . ...... . . . . . . ... . . . . . . ... . . . ........................ . . . . . . ... . . . . . .... ......... . . . ...... . . .......                                                                                                                                     . . . . . . .\n',
 '000021000021 806.84601169505000 806.84601169505000 454.86041741251200 454.8604174125120020362014. 222 91.3125121.7500 91.3125

Taking a look at the main fraim file, code was from the example given above.

In [114]:
lines_list = []
files = ['assign_wk3/nispuf14.dat']
for file in files:
    with open(file) as infile:
        for line in infile:
            if len(line) > 1:    # Blank lines at the end of files.
                lines_list.append(line.strip()) # strip() leaves empty blank lines -- skip these

In [115]:
lines_list[:3]

['000011000012    .                  .               218.30024855484000 218.3002485548400010222014. 223365.2500152.1875182.6250..3 2 131 2 .4223.000000000000001142131299999912 12 222212..0..........................................................................................   .   .   .   .   .   ....   .   .   .   .   .   ....   .   .   .   ......   .   .   .   .   .  .  ...   .   .   .   .   .  .  ...   .   .  .  ...............  .........   .   .   .   .   .  .  ...   .   .   .   .  .  ....  .........  .  .   .  .  .....   .   .   ....... . . . . . .... . . . . . .... . . . ...... . . . . . . ... . . . . . . ... . . . ........................ . . . . . . ... . . . . . .... ......... . . . ...... . . .......                                                                                                                                     . . . . . . .',
 '000021000021 806.84601169505000 806.84601169505000 454.86041741251200 454.8604174125120020362014. 222 91.3125121.7500 91.3125..

In [116]:
len(lines_list)

24897

In [117]:
print(lines_list[0])

000011000012    .                  .               218.30024855484000 218.3002485548400010222014. 223365.2500152.1875182.6250..3 2 131 2 .4223.000000000000001142131299999912 12 222212..0..........................................................................................   .   .   .   .   .   ....   .   .   .   .   .   ....   .   .   .   ......   .   .   .   .   .  .  ...   .   .   .   .   .  .  ...   .   .  .  ...............  .........   .   .   .   .   .  .  ...   .   .   .   .  .  ....  .........  .  .   .  .  .....   .   .   ....... . . . . . .... . . . . . .... . . . ...... . . . . . . ... . . . . . . ... . . . ........................ . . . . . . ... . . . . . .... ......... . . . ...... . . .......                                                                                                                                     . . . . . . .


In [118]:
print(len(lines_list[0]))

866


It looks like there are 24,897 records, each of which as 866 lines long. This makes sense as the final comma insertion ends at lie 866

Now I need a way to insert commas into the main frame, I found a simple method for doing so on stack over flow

In [119]:
#https://stackoverflow.com/questions/5254445/how-to-add-a-string-in-a-certain-position
def insertChar(mystring, position, chartoinsert ):
    mystring   =  mystring[:position] + chartoinsert + mystring[position:] 
    return mystring 

Now to pull everythign together, I need to iterate through the entire mainframe, add commas at the correct position, and then return those edited rows to a new list.

In [120]:
# Method for adding commas at the insertion points
# I used a lot of print commands to figure out how to get this to work, I left them in the code to show my work
i = 0 # Value for the mainframe row counter

row_list = []

slice_spot = results

while i < 24897: #This needs to be change, just making it work on one element at a time
    x = (lines_list[i])
    # print("THIS IS X", x)
    # print("THIS IS WHAT TYPE X IS", type(x))
    # print("THIS IS I",i)
    i = i+1
    j = 0
    k = 0
    while j < 922:
        stop = int(slice_spot[j+1])
        # print(stop)
        location = int(stop + k)
        # print("Location", location)
        x = insertChar(x,location, ',')
        # print("This is the new x", x)
        j = j+2
        # print("THIS IS J", j)
        k = k+1
        if j == 922:
            x = insertChar(x,location+1,'end')
            # print("THIS IS THE FINAL X", x)
            # print("NOW THIS RAN")
            # print("THIS IS THE FORMAT OF X", type(x))
            row_list += [x]
        

Making sure that that worked

In [121]:
print(row_list)

IOPub data rate exceeded.
The notebook server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--NotebookApp.iopub_data_rate_limit`.

Current values:
NotebookApp.iopub_data_rate_limit=1000000.0 (bytes/sec)
NotebookApp.rate_limit_window=3.0 (secs)



In [65]:
joined_rows = []
row = '' # Just creates the variable

for line in (row_list):
    row += line + ','
    if 'end' in line:
        joined_rows.append(row[:-1])
        row = ''

In [122]:
# checking that all the rows are present
len(joined_rows)

24897

Taking a look at a random row

In [123]:
print(joined_rows[42])

000401,00040,1,1345.02638090789000,1345.02638090789000, 486.66722751075900, 486.66722751075900,2040,2014,2, 1,2,3,273.9375,152.1875,213.0625,.,.,6, 1, 1,2,3, 1, 2,3,1,2,0.96376252891288,3, 8,.,1,2,1,2, ., 2, 2,1,2, 1,1, 40,40,27,2,1,1,2,4,4,1,.,1,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,3,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,   .,   .,   .,   .,   .,   .,.,.,.,   .,   .,   .,   .,   .,   .,.,.,.,   .,   .,   .,   .,.,.,.,.,.,   0,   .,   .,   .,   .,  .,  .,.,.,   .,   .,   .,   .,   .,  .,  .,.,.,   .,   .,  .,  .,.,.,.,.,.,.,.,.,.,.,.,.,.,.,  .,.,.,.,.,.,.,.,.,   .,   .,   .,   .,   .,  .,  .,.,.,   .,   .,   .,   .,  .,  .,.,.,.,  .,.,.,.,.,.,.,.,.,  .,  .,   .,  .,  .,.,.,.,.,   .,   .,   .,.,.,.,.,.,., ., ., ., ., ., .,.,.,., ., ., ., ., ., .,.,.,., ., ., ., .,.,.,.,.,., 0, ., ., ., ., ., .,.,., ., ., ., ., ., ., .,.,., ., ., ., .,.,.,.,.,.,.,.,.,.,.,.,.,.,.,.,.,.,.,.,.,.,.,., ., ., ., ., ., .,

In [124]:
# Making sure the format is correct
type(joined_rows)

list

I can now dump everything into a json file using the method provided in class

In [125]:
rows = [row.split(',') for row in joined_rows]
complete_json = [dict(zip(headers,row)) for row in rows]

Making sure that that worked correctly

In [126]:
complete_json[1]

{'SEQNUMC': '000021',
 'SEQNUMHH': '00002',
 'PDAT': '1',
 'PROVWT_D': ' 806.84601169505000',
 'PROVWT_D_TERR': ' 806.84601169505000',
 'RDDWT_D': ' 454.86041741251200',
 'RDDWT_D_TERR': ' 454.86041741251200',
 'STRATUM': '2036',
 'YEAR': '2014',
 'AGECPOXR': '.',
 'HAD_CPOX': ' 2',
 'SHOTCARD': '2',
 'AGEGRP': '2',
 'BF_ENDR06': ' 91.3125',
 'BF_EXCLR06': '121.7500',
 'BF_FORMR08': ' 91.3125',
 'BFENDFL06': '.',
 'BFFORMFL06': '.',
 'C1R': '6',
 'C5R': ' 1',
 'CBF_01': ' 1',
 'CEN_REG': '2',
 'CHILDNM': '3',
 'CWIC_01': ' 2',
 'CWIC_02': ' .',
 'EDUC1': '2',
 'FRSTBRN': '1',
 'I_HISP_K': '2',
 'INCPORAR': '0.50000000000000',
 'INCPOV1': '3',
 'INCQ298A': ' 4',
 'INTRP': '.',
 'LANGUAGE': '1',
 'M_AGEGRP': '2',
 'MARITAL2': '1',
 'MOBIL_I': '1',
 'NUM_PHONE': ' .',
 'NUM_CELLS_HH': ' 2',
 'NUM_CELLS_PARENTS': ' 2',
 'RACE_K': '1',
 'RACEETHK': '2',
 'RENT_OWN': ' 2',
 'SEX': '1',
 'ESTIAP14': ' 36',
 'EST_GRANT': '36',
 'STATE': '18',
 'D6R': '1',
 'D7': '1',
 'N_PRVR': '1',
 'PROV_FAC

Now to export the JSON file

In [127]:
import json

with open("assign_wk3/completed_json.json",'w') as outfile:
    json.dump(complete_json, outfile)

# Conclusion
#### Main frame data is often times saved in character spaced files, meaning that the values are saved based on their characters position wihtout anytype of easily readable spacing. While this makes for smaller storage space it does create a problem when the data needs to be pulled out.

#### This notebook demonstrates a method for pulling the data frame information out of a PDF and then turning it into a JSON file based on the information provided in the PDF.
